### ISSUE TITLE
Cilium agent fails (timeout) to sync kubernetes cache after restart

### ISSUE THREAD

**Issue by cehoffman** [cilium/cilium#9697]
## Support

I have cilium agent deployed through the 1.7.0-rc1 helm chart, the config is at the end. The first time the agent started on each of the nodes (with workload already there because flannel was providing the vxlan fabric) everything was fine. I rebooted the nodes to switch to complete cilium management of the network. Again everything is fine. When cilium restarts though on a node, it never becomes ready. It stops at `level=info msg="Waiting until all pre-existing resources related to policy have been received" subsys=k8s-watcher`. It eventually times out waiting for this condition, exits and then is restarted.

I'm now trying this on the best of hardware, but the default 3 minutes should have been plenty of time to sync and it does work perfectly on a fresh reboot of the node. To explore the timeout more I pulled the code and extended the timeout to 10 minutes from the hardcoded 3 and made corresponding changes to the liveness probe in the daemonset to let it persist long enough for that timeout.

Again, the process never finished syncing. I then enabled debug to see if any of the individual resource type cache finished sync log lines occurred. None appears, it seems the first informer for network policies never completes which is should be taxing. There are only 2 networkpolicies in this cluster. The full logs from a container with debug on in this case is [here](https://gist.github.com/ed68eaa7146e79f317aaa7669ed72536).

Since the process connects to the apiserver earlier, I don't think it is a reachability issue although it definitely feels like it. I do have the non cluster network host and port of the apiserver set in the daemonset for the agent. It is not trying to use the internal cluster service address.

For a bit of context, this cluster is very small and has nearly no work occurring in it but it has the exact same configuration as our much larger cluster installs. I'm using it as a testing ground for new infrastructure services and components. Thus I know other informer based caches do work after restart, for example cert-manager and controller-manager itself.

```yaml
  auto-direct-node-routes: "false"
  bpf-ct-global-any-max: "262144"
  bpf-ct-global-tcp-max: "524288"
  cluster-name: home
  cni-chaining-mode: portmap
  container-runtime: none
  debug: "false"
  enable-host-reachable-services: "true"
  enable-ipv4: "true"
  enable-ipv6: "false"
  enable-node-port: "true"
  host-reachable-services-protos: udp tcp
  identity-allocation-mode: crd
  install-iptables-rules: "true"
  k8s-require-ipv4-pod-cidr: "true"
  masquerade: "true"
  monitor-aggregation: medium
  preallocate-bpf-maps: "false"
  sidecar-istio-proxy-image: cilium/istio_proxy
  sockops-enable: "false"
  tofqdns-enable-poller: "false"
  tunnel: vxlan
  wait-bpf-mount: "true"
```

Edit:

Updated gist to one with timestamps for each line

**Comment by cehoffman** [comment]
After posting this I decided to try removing the overrides for apiserver host and port. The resulted in the pods all becoming ready, but that has the issue on restart the pods can't reach the apiserver because there has not been a route for the in cluster service setup yet.

**Comment by cehoffman** [comment]
Dug a bit more into this any my current opinion is that after restart during  a reconfigure the opened and working connection to the apiserver is somehow broken when using the public interface the points to an external load balancer for the apiservers. If I set the kubernetes host to be the kubernetes.default.svc domain, all works on cilium restart because that service endpoints use the private internal ip addresses of each apiserver. If I set the `advertise-ip` flag on the apiserver to be the public load balancer then restart fail when using kubernetes.default.svc which matches the starting behavior.

If I change the kubernetes hostname to be one of the apiserver host names, restarts work. This would appear to show that DNS not at fault, but it is perplexing because the public IP on the same segment as the private IPs.

More clearly, each of the apiserver hosts has a single private IP of

- 192.168.1.175
- 192.168.1.176
- 192.168.1.177

The load balancer is uses IP 192.168.1.253. The load balancer is on a separate host and not part of the kubernetes system. It is pfsense actually.

**Comment by brb** [comment]
Thanks for the issue. A few questions below:

1.Have you tried setting `KUBERNETES_SERVICE_HOST` to `192.168.1.253` instead of `k8s.hsv-a.ceh.im`?
2. Have you tried running in a clean environment, i.e. without flannel being previously installed?
3. Can you record pcaps with tcpdump for `cilium-agent -> LB` and `LB -> kube-apiserver`?

**Comment by cehoffman** [comment]
Thanks for taking the time with this.

1. I had not tried straight IP, but on doing so the behavior is the same as using the DNS. The first iteration actually failed at TLS setup because I didn't have the IP in the certificate, but after resolving that it timeouts on waiting for resource sync.
2. I haven't tried from a completely from the ground up kubernetes, but all the nodes have been cycled to clean state. The only remnants left are the annotations on the resources from flannel, e.g. `flannel.alpha.coreos.com/backend-data: '{"VtepMAC":"fa:c9:c6:35:ee:05"}'` on the node resource
3. I was able to get from cilium-agent->LB 
[cilium-to-lb.pcap.zip](https://github.com/cilium/cilium/files/3914944/cilium-to-lb.pcap.zip) Getting from the LB to apiserver may be a bit more challenging to keep it content only related to this traffic. I'll work on that.

**Comment by cehoffman** [comment]
I was attempting to get the LB->apiserver traffic in an isolated manner, and think the problem may lie with the LB in pfsense that I'm using. I created a simple haproxy instance on my laptop to front the apiservers and added my laptop IP as an identity to the apiserver certificate. I verified the proxy was working and then configured cilium to use it. Well everything worked this time, so I guess I'll look into a different LB setup.
